### DISCUSSION TITLE
Geocode externally

### DISCUSSION THREAD

**Discussion by AneurinPartridge** [alexreisner/geocoder#1579] [Ideas]
I'm trying to do my geocoding externally as the api I'm using isn't supported by the geocoder gem, however I want to access the near method to do the location comparison, I'm able to geocode but I believe I need to use geocoded_by to access it.
The api I'm using is called getAddress.io and I want to use that to geocode but use comparison to existing addresses that are already geocoded.
Apologies for any misunderstanding, I'm new to coding and trying my best :)

**Comment by AneurinPartridge** [comment]
I've circumvented this by putting 
```
geocoded_by :geocode_space_filler

  def geocode_space_filler
    response = RestClient.get("https://api.getAddress.io/find/#{post_code.gsub(/\s+/, "") }?api-key=#{ENV.fetch("GETADDRESS_API_KEY")}")
    lat = JSON.parse(response)["latitude"]
    long = JSON.parse(response)["longitude"]
    if !lat || !long
      puts "could not find postcode"
      return
    end
    self.update(latitude: lat, longitude: long)
  end
```

**Comment by alexreisner** [comment]
Yes, that's a valid workaround. Something simpler should also work. For example, `geocoded_by :post_code` should be sufficient. The gem won't actually try to do any geocoding unless you ask it to--the `geocoded_by` line is mainly a way of telling the gem that the class stores lat/long data and that the `near` method should be available.

**Comment by AneurinPartridge** [comment]
Hi @alexreisner , at the risk of annoying you further(completely unintentional) but just for some help, and I don't want to bloat issues/ discussions with more of my questions, is there a way I can save the geoapify api response so I can query it in the database and use an api call as a fallback?

**Comment by alexreisner** [comment]
Not annoying at all, this is an interesting use case. But I don't quite understand what you're trying to achieve. You're using two APIs, Getaddress.io (not supported by Geocoder) and Geoapify (supported), and you want to call one or the other based on certain conditions? This is how you could call Getaddress.io if Geoapify fails:

    geocoded_by :address # or wherever your address string is stored
    after_validation :geocode_with_fallback

    def geocode_with_fallback
      geocode # this kicks off an API call via Geocoder
      if not geocoded?
        # call Getaddress.io
      end
    end

but I don't think that's what you're trying to do.

**Reply by AneurinPartridge** [reply]
I'm using getaddress to do the geocoding of specific addresses and all autofill etc.
I'm using geoapify to geocode a town/city so i'm able to run near on the coordinates with the lat/long of the addresses I have stored.
I want to be able to reduce api calls to geoapify by caching responses for the town/city calls so it speeds up query time.
Sorry if I'm wording this poorly, basically, the only geocoding with geoapify I'm doing is town/city as getaddress does not actually return coordinates for these, so I'm being forced to use an alternative.
I appreciate you taking the time to help me with this and I'm glad you're so helpful :)
`Course.where(address_id: Address.near(location_params[:place], location_params[:radius].to_i).map(&:id))`
This is the code that uses geoapify, everything else I can do with getaddress.

**Comment by AneurinPartridge** [comment]
Hi Alex, I'm just leaving a note that someone else will be taking over trying to work on this called Lewys, sorry to be a pain

**Comment by alexreisner** [comment]
So, Geocoder does have a built-in caching feature that should eliminate duplicate calls to Geoapify. If that would solve the problem, see this section of the README: https://github.com/alexreisner/geocoder#caching
